toolchain/base/value_store.h

```diff
@@ -28,6 +29,12 @@ namespace Internal {
 class ValueStoreNotPrintable {};
 }  // namespace Internal
 
+#ifndef CARBON_POISON_VALUE_STORES
+#define CARBON_POISON_VALUE_STORES 0
+#elif !LLVM_ADDRESS_SANITIZER_BUILD
+#error "CARBON_POISON_VALUE_STORES requires address sanitizer"
+#endif
```

Comment:
I'd add some documentation here. The bazel flag seems awfully distant from this code, so some cross-connecting docs / TODOs seem helpful.

```suggestion
// Setup our compile time condition controlling poisoning of value stores. This is set to one by the Bazel flag `--features=poison_value_stores`.
//
// TODO: Eventually, this will always enabled when ASan is enabled, but we can't do that until we clean up all of the latent bugs.
#ifndef CARBON_POISON_VALUE_STORES
#define CARBON_POISON_VALUE_STORES 0
#elif !LLVM_ADDRESS_SANITIZER_BUILD
#error "CARBON_POISON_VALUE_STORES requires address sanitizer"
#endif
```